In [1]:
# initial setup
%run "../../../common/0_notebooks_base_setup.py"


/media/paulati/Nuevo vol/paula/dh/2021/dsad_2021_paula/common
default checking
Running command `conda list`... ok
jupyterlab=2.2.6 already installed
pandas=1.1.5 already installed
bokeh=2.2.3 already installed
seaborn=0.11.0 already installed
matplotlib=3.3.2 already installed
ipywidgets=7.5.1 already installed
pytest=6.2.1 already installed
chardet=4.0.0 already installed
psutil=5.7.2 already installed
scipy=1.5.2 already installed
statsmodels=0.12.1 already installed
scikit-learn=0.23.2 already installed
xlrd=2.0.1 already installed
Running command `conda install --yes nltk=3.5.0`... ok
Solving environment: ...working... done

# All requested packages already installed.


unidecode=1.1.1 already installed
pydotplus=2.0.2 already installed
pandas-datareader=0.9.0 already installed
flask=1.1.2 already installed


In [2]:
import pandas as pd
import numpy as np

<link rel="stylesheet" href="../../../common/dhds.css">
<div class="Table">
    <div class="Row">
        <div class="Cell grey left"> <img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_portada.jpg" align="center" width="90%"/></div>
        <div class="Cell right">
            <div class="div-logo"><img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/common/logo_DH.png" align="center" width=70% /></div>
            <div class="div-curso">DATA SCIENCE</div>
            <div class="div-modulo">MÓDULO 4</div>
            <div class="div-contenido">Regresión Logística</div>
        </div>
    </div>
</div>

## Agenda

---

- Describir el modelo de regresión logística

- Presentar los fundamentos matemáticos de la regresión logística

- Entrenar modelos de regresión logística utilizando la biblioteca Scikit-learn





<img src = "https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_logisticregression.png" align="center" />


<div class="div-dhds-fondo-1"> Introducción
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Introducción

---


La regresión logística es un abordaje lineal para resolver **problemas de clasificación**.

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo1.png" align="center" />


¿Podríamos usar una regresión lineal?

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo1_lineal.png" align="center" />


Podríamos, pero vamos a obtener valores fuera del rango [0,1], dificultando la interpretación como probabilidad.

La regresión logística nos permite **modelar la probabilidad** de que la variable objetivo `y` pertenezca a una determinada categoría, dados los valores de las variables `X`.   

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo1_logistica.png" align="center" />



Debemos establecer una **frontera de decisión lineal**

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo1_logistica_umbral.png" align="center" />



## Introducción (cont.)

---

Veamos ahora un ejemplo con dos features

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo2.png" align="left" width ="50%"/>


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo2_3D.png" align="right" width ="50%" />




También podemos definir una **frontera de decisión lineal** que resuelva el problema

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo2_umbral.png" align="center"/>



## Ejemplo Regresión Logística

---


Queremos predecir la **probabilidad** de que un cliente no pague su tarjeta de crédito (entre en Default):

$$ y = \cases{
    0, & \text{si  No}\\
    1, & \text{si  Sí}
}$$

Los **features** son:

- Ingresos (income)

- Deuda en su tarjeta de crédito (balance)



<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo3.png" align="center"/>

Observando los datos, decidimos usar como variable predictora `balance` (¿por qué?). 

Es decir, queremos predecir **p(y = 1 | balance)**

Definiendo como umbral 0.5 (podríamos elegir otro!), si la p(y=1|balance) > 0.5 => default = Sí



<div class="div-dhds-fondo-1"> Regresión Logística
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## ¿Por qué no podemos estimar la probablidad usando una regresión lineal simple?

---

Si estimáramos p(y=1 | X) con una regresión lineal, nuestro modelo asumiría la siguiente forma:

<p style="font-size:20px;">
$$p(X) = \beta_0 + \beta_1 . X_1$$
</p>
donde para abreviar, definimos 
<p style="font-size:20px;">
$$p(X) = p(y = 1 | X)$$
</p>

Problemas:

- Esto arrojaría **valores fuera del rango válido para una probabilidad [0,1]**. 

- Cuando el problema de clasificación es multiclase, el modelo **interpretaría a las diferentes clases como valores numéricos**.



<p style="font-size:20px;">
Tenemos que buscar una función que nos garantice que las estimaciones que hagamos estarán dentro del rango válido de una probabilidad.
</p>

## La función logística

---

<p style="font-size:24px;">
$$p(X) = \frac{e^{\beta_0 + \beta_1 . X}}{1 + e^{\beta_0 + \beta_1 . X}}$$
</p>

Sin importar qué valores tome `X` siempre vamos a predecir valores dentro del rango [0,1].


## Lineal vs. Logística

---

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo3_probability.png" align="center" />

## odds-ratio

---

El odd es la **probabilidad de que suceda un evento dividido por la probabilidad de que no suceda**. 

No hay una traducción clara al español, así que usamos el término odd. 

Los odd **oscilan entre 0 e infinito** y se pueden calcular para la ocurrencia del evento como para la no ocurrencia del evento. 

<p style="font-size:20px;">
$$\frac{p(X)}{1- p(X)} = e^{\beta_0 + \beta_1 . X}$$
</p>

**Ejemplo**:

El odd de sí-votar al partido A es 0.35 / 0.65 = 0.53

El odd de no-votar al partido A es 0.65 / 0.35 = 1.86

Entonces es 1.86 veces más probable que alguien no sea del partido A a que lo sea. 

**Los odds se interpretan como ratios, es decir, la cantidad de veces que algo pueda suceder sobre que no pueda suceder**.


## log odds

---


El logaritmo de odds-ratio tiene una relación lineal con $X$.

<p style="font-size:20px;">
$$ log(\frac{p(X)}{1 - p(X)}) = \beta_0 + \beta_1 . X$$
</p>

En el modelo lineal, los $\beta_p$ son el cambio promedio en $y$ ante un cambio unitario en $X_p$. 

En la regresión logística, incrementar una unidad en $X$, cambia el logaritmo del odds-ratio en $\beta_1$. 

O, lo que es lo mismo, multiplica el odds por $\beta_1$. 

La relación entre $p(X)$ y $X$ no es una línea recta, entonces cuánto cambia $p(X)$ ante un cambio unitario en $X$ depende de los valores de $X$. 

El signo de $\beta_1$ expresa la dirección de cambio en $p(X)$, independientemente del valor de $X$.

## Probabilidad, odds-ratio, log odds

---


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_odds.png" align="center" />

<div class="div-dhds-fondo-1"> Ejemplos
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Ejemplo Regresión Logística. Ajuste variable predictora cuantitativa.

---


Siguiendo nuestro ejemplo de default en tarjetas de crédito obtenemos estos valores para el modelo ajustado:
    
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo3_fit1.png" align="center"/>

Observamos que un incremento de $1 en el balance incrementa 0.0055 unidades el "log odds-ratio".

Existe un estadístico z que es análogo al estadístico t en regresión lineal. 

$H_0$: $\beta_1 = 0$ (la probabilidad de default no depende del balance)

$H_1$: $\beta_1 \ne 0$


## Ejemplo Regresión Logística - Predicción

---

Una vez que hemos estimados los coeficientes del modelo, podemos hacer predicciones y computar la probabilidad de default para algún valor dado de balance.

Por ejemplo, para un balance \$1000 tenemos que está por debajo del 1%

<p style="font-size:20px;">
$$ \hat{p}(X) = \frac{e^{\hat{\beta_0} + \hat{\beta_1} . X}}{1 + e^{\hat{\beta_0} + \hat{\beta_1} . X}} = \frac{e^{-10.6513 + 0.0055 . 1000}}{1 + e^{-10.6513 + 0.0055 . 1000}} = 0.00576 $$
</p>

En cambio, para un balance de \$2000 es mucho más grande y está cerca del 58%

<p style="font-size:20px;">
$$ \hat{p}(X) = \frac{e^{-10.6513 + 0.0055 . 2000}}{1 + e^{-10.6513 + 0.0055 . 2000}} = 0.586 $$
</p>


## Ejemplo Regresión Logística. Ajuste variable predictora cualitativa.

---

La regresión logística admite también variables cualitativas. 

Entrenamos un modelo utilizando la condición de ser estudiante sobre la probabilidad de default.

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo3_fit2.png" align="center" /> 

En este caso, el coeficiente es positivo, lo cual indica que **ser estudiante tiene una relación positiva con ser potencial moroso**. 

¿Qué pueden decir de la significancia del coeficiente?




## Ejemplo Regresión Logística - Predicción

---

Podemos estimar las probabilidades de entrar en default siendo estudiante y no siéndolo

<p style="font-size:20px;">
$$ \hat{p}(default = Sí | X = 1) = \frac{e^{-3.5041 + 0.4049 . 1}}{1 + e^{-3.5041 + 0.4049 . 1}} = 0.0431 $$    
</p>

<p style="font-size:20px;">
$$ \hat{p}(default = Sí | X = 0) = \frac{e^{-3.5041 + 0.4049 . 0}}{1 + e^{-3.5041 + 0.4049 . 0}} = 0.0292 $$    
</p>



<div class="div-dhds-fondo-1"> Regresión Logística Múltiple
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Regresión Logística Múltiple

---

De forma análoga al caso de regresión lineal, pensemos en el problema de predecir una variable cualitativa binaria con una serie de p features.

Nuestro modelo de regresión logística múltiple quedará definido:

<p style="font-size:24px;">
$$p(X) = \frac{e^{\beta_0 + \beta_1 . X_1 + \beta_2 . X_2 + ... + \beta_p . X_p}}{1 + e^{\beta_0 + \beta_1 . X_1 + \beta_2 . X_2 + ... + \beta_p . X_p}}$$
</p>


Y reescribiéndolo en términos de logs odds:

<p style="font-size:20px;">
$$ log(\frac{p(X)}{1 - p(X)}) = \beta_0 + \beta_1 . X_1 + \beta_2 . X_2 + ... + \beta_p . X_p$$
</p>




## Ejemplo Regresión Logística Múltiple

---

Veamos los resultados de aplicar este modelo para predecir la probabilidad de default según el ingreso, el balance y la condición de estudiante.

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo3_fit3.png" align="center" />

¿Qué pueden decir de los resultados?

¿Ven algo raro?


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_ejemplo4.png" align="center" />

En el cuadro de la izquierda, en <a style="color:orange;">naranja</a> se representan a los estudiantes y en <a style="color:blue;">azul</a> a los no estudiantes. 

La línea punteada representa el % total de defaults, mientras que la línea sólida representa el % de defaults en función del balance de la tarjeta de crédito


<div class="div-dhds-fondo-1"> Función de costo
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Función de costo

---

Como en todos los algoritmos de Machine Learning, debe definirse una **función de costo que dependa de los parámetros del modelo**. De esta manera, el entrenamiento o "aprendizaje" consiste en **encontrar los parámetros que optimicen dicha función**.

En el caso de la regresión logística se busca **maximizar la verosimilitud del modelo**, es decir que las estimaciones de las probabilidades de las observaciones que pertenezcan a la clase 1 sean cercanas a 1 y lo mismo para las que pertenezcan a la clase 0. 

Matemáticamente:

<p style="font-size:24px;">
$$L = \prod_{i: y_i = 1}P(y = 1 | x = X) . \prod_{i: y_i = 0}(1 - P(y = 1 | x = X))$$
</p>    

Tomando el logaritmo:
    
<p style="font-size:24px;">
$$LL = \sum_{i=1}^m{y_i . log(P(y_i = 1 | x_i = X)) + (1-y_i).log(1 - P(y_i = 1 | x_i = X))}$$
</p>    


<p style="font-size:24px;">
$$LL = \sum_{i=1}^m{y_i . log(P(y_i = 1 | x_i = X)) + (1-y_i).log(1 - P(y_i = 1 | x_i = X))}$$
</p>    

Cuando $y_i = 1$ se suman los:

<p style="font-size:24px;">
$$log(P(y_i = 1 | x_i = X))$$
</p>    

que deberían resultar en valores cercanos a 1.

Cuando $y_i = 0$ se suman los:

<p style="font-size:24px;">
$$log(1 - P(y_i = 1 | x_i = X))$$
</p>    

que deberían resultar en valores cercanos a 0.

Esto se conoce como **Maximum Likelihood Estimation o MLE**. 

Como la función de costo es algo que debe minimizarse, se trabaja con el opuesto y se promedia:

<p style="font-size:24px;">
$$J = \frac{-1}{m} . LL$$
</p>    


##  Regresión Logística con una variable predictora

---

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_logistic_animation.gif" align="center" width="70%" />

##  Regresión Logística con dos variables predictoras

---

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_logistic_animation_2.gif" align="center" width="60%"/>

<div class="div-dhds-fondo-1"> Regularización
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Regularización

---

<table>
<tr>    
<td style="width:60%;font-size:16px;line-height:2;">    
Recordemos que:
<ul>    
<li>Podemos pensar en la regularización como agregar (o aumentar) el sesgo si nuestro modelo sufre de (alta) varianza (es decir, sobreajusta los datos de entrenamiento).</li> 
<li>Por otro lado, un sesgo excesivo dará como resultado un ajuste insuficiente (un indicador característico de un sesgo alto es que el modelo muestra un rendimiento "malo" tanto para el conjunto de datos de entrenamiento como de prueba).</li> 
<li>Sabemos que nuestro objetivo en un modelo no regularizado es minimizar la función de costo, es decir, queremos encontrar los pesos de las features que corresponden al costo mínimo global (la función de costo en la regresión logística es convexa).</li>     
</ul>
</td>
<td>
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_unregularized.png" align="center"/>
</td>    
</tr>    
</table>

<table>
<tr>    
<td style="width:60%;font-size:16px;line-height:2;">    


Si regularizamos la función de costo (por ejemplo, a través de la regularización L2), agregamos un término adicional a nuestra función de costo (J) que aumenta a medida que aumenta el valor de los pesos de sus parámetros (w ó $\beta$). 

    
Tengamos en cuenta que a la regularización agregamos un nuevo hiperparámetro, $\lambda$, para controlar la fuerza de la regularización.

        
<p style="font-size:20px;">
$L2: \frac{\lambda}{2} . \lVert{w}\rVert^2  = \frac{\lambda}{2} . \sum_{j=1}^{p}w_j^2$
</p>

<p style="font-size:20px;">
$J(w) = J + \frac{\lambda}{2} . \lVert{w}\rVert^2 $
</p>    

</td>
<td>
        
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_regularized.png" align="center"/>
    
</td>    
</tr>    
</table>    

<div class="div-dhds-fondo-1"> Scikit-learn
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Ejemplo 

---

Vamos a ajustar un modelo de regresión logística para el dataset <a href="https://rdrr.io/cran/ISLR/man/Default.html" taget="_blank">Default</a>, un conjunto de datos simulados que contiene información sobre diez mil clientes. 

El objetivo aquí es predecir qué clientes incumplirán con la deuda de su tarjeta de crédito.

Vamos a usar scikit-learn <a href="https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html" target="_blank">LogisticRegression</a>.

Esta clase implementa la regresión logística regularizada utilizando la biblioteca "liblinear", los métodos "newton-cg", "sag", "saga" y "lbfgs". **Tengamos en cuenta que la regularización se aplica por default**.

In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix

In [4]:
data = pd.read_csv('../Data/Default.csv', sep="\t")
data.head()

default student      balance        income
0      No      No   729.526495  44361.625074
1      No     Yes   817.180407  12106.134700
2      No      No  1073.549164  31767.138947
3      No      No   529.250605  35704.493935
4      No      No   785.655883  38463.495879

Vemos que las categorías están muy desbalanceadas:

In [5]:
data.default.value_counts()

No     9667
Yes     333
Name: default, dtype: int64

Convertimos a dummies las variables categóricas:

In [6]:
student_dummy = pd.get_dummies(data.student, drop_first = True, prefix = 'student')
student_dummy.head(3)

student_Yes
0            0
1            1
2            0

In [7]:
data = pd.concat([data, student_dummy], axis = 1)
data.head(3)

default student      balance        income  student_Yes
0      No      No   729.526495  44361.625074            0
1      No     Yes   817.180407  12106.134700            1
2      No      No  1073.549164  31767.138947            0

In [8]:
X = data[['student_Yes', 'balance', 'income']]
y = data.default

Creamos los conjuntos de entrenamiento y test, estableciendo el valor del parámetro `stratify` para que los datos se dividan de forma estratificada según las etiquetas de default. Esto construye conjuntos de entrenamiento y test que mantienen la misma proporción de registros en cada categoría que en el dataset original.

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state = 12)

Comprobemos que se mantienen las proporciones:

In [10]:
data.default.value_counts()['Yes'] / data.shape[0]

0.0333

In [11]:
y_train.value_counts()['Yes'] / y_train.shape[0]

0.03333333333333333

In [12]:
y_test.value_counts()['Yes'] / y_test.shape[0]

0.0332

Instanciamos una regresión logística:

In [13]:
logistic_regression = LogisticRegression(penalty='none') # sin regularización
logistic_regression.fit(X_train, y_train);

In [14]:
y_test_pred = logistic_regression.predict(X_test)
y_test_pred

array(['No', 'No', 'No', ..., 'No', 'No', 'No'], dtype=object)

In [15]:
y_test_pred_proba = logistic_regression.predict_proba(X_test)
y_test_pred_proba

array([[9.97675832e-01, 2.32416770e-03],
       [9.98876259e-01, 1.12374110e-03],
       [9.96638594e-01, 3.36140624e-03],
       ...,
       [9.67782530e-01, 3.22174700e-02],
       [9.14285262e-01, 8.57147383e-02],
       [9.99920761e-01, 7.92386434e-05]])

In [16]:
accuracy_score(y_test, y_test_pred)

0.9684

Aunque el valor de accuracy resultó alto (97%), vemos en la matriz de confusión que la performance para la etiqueta 'Yes' es bastante mala.

De los 83 registros con etiqueta 'Yes'

- clasificó correctamente 17, o sea 20%.

- clasificó erroneamente 66, o sea 80%.

Para la etiqueta 'No' la performance es buena:

De los 2417 registros con etiqueta 'No'

- clasificó correctamente 2404, o sea 99.5%.

- clasificó erroneamente 13, o sea 0.5%.



In [17]:
confusion_matrix(y_test, y_test_pred)

array([[2404,   13],
       [  66,   17]])

El valor de $\beta_0$ en la regresión logística es

In [18]:
logistic_regression.intercept_

array([-3.01195506])

Los $\beta$ de la regresión logística son:

In [19]:
logistic_regression.coef_

array([[-4.01670288e+00,  4.10753666e-03, -1.31786584e-04]])

El modelo queda representado por
<p style="font-size:24px;">
$$ \hat{p}(X) = \frac{e^{\hat{\beta_0} + \hat{\beta_1} . student\_Yes + \hat{\beta_2} . balance + \hat{\beta_3} . income}}{1 + e^{\hat{\beta_0} + \hat{\beta_1} . student\_Yes + \hat{\beta_2} . balance + \hat{\beta_3} . income}} = \frac{e^{-3.01 - 4.01 . student\_Yes + 0.004 . balance - 0.0001 . income}}{1 + e^{-3.01 - 4.01 . student\_Yes + 0.004 . balance - 0.0001 . income}}$$
</p>

Observaciones:

**No se requiere estandarización para la regresión logística**. El objetivo principal de estandarizar features es ayudar a la convergencia de la técnica utilizada para la optimización. 

Si usamos **regresión logística con regularización de Lasso o Ridge debemos estandarizar las features**

En la práctica guiada verán un ejemplo de uso de statsmodels para ajustar una regresión logística.


<div class="div-dhds-fondo-1"> Hands-on
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Hands-on

---

### Ejercicio 1

Aplicar regularización de Lasso para predecir la variable default del dataset Default.csv


### Ejercicio 2

2.1 Asignar aleatoriamente la clase 'Unknown' al 20% de los registros que actualmente pertenecen a la clase 'No'.

2.2 Ajustar un modelo de regresión logística multiclase usando scikit-learn

2.3 Construir un dataset con las tres clases balanceadas y ajustar un modelo de regresión logística multiclase usando scikit-learn. Comparemos los resultados con el modelo anterior.



## Solución

---

### Ejercicio 1

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html


In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, multilabel_confusion_matrix
from sklearn.preprocessing import StandardScaler

In [21]:
data = pd.read_csv('../Data/Default.csv', sep="\t")

student_dummy = pd.get_dummies(data.student, drop_first = True, prefix = 'student')
student_dummy.head(3)

data = pd.concat([data, student_dummy], axis = 1)

X = data[['student_Yes', 'balance', 'income']]
y = data.default

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state = 12)

Como aplicamos regularización, estandarizamos las features:

In [23]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [24]:
logistic_regression = LogisticRegression(penalty='l1', solver= 'liblinear')
logistic_regression.fit(X_train_scaled, y_train);

In [25]:
y_test_pred = logistic_regression.predict(X_test_scaled)
y_test_pred

array(['No', 'No', 'No', ..., 'No', 'No', 'No'], dtype=object)

In [26]:
accuracy_score(y_test, y_test_pred)

0.9744

In [27]:
confusion_matrix(y_test, y_test_pred)

array([[2404,   13],
       [  51,   32]])

La matriz de confusión es mejor que la obtenida en el ejemplo sin regularización (hay menos casos mal clasificados).


In [28]:
logistic_regression.coef_

array([[-0.29599402,  2.72356118,  0.09481735]])

### Tarea (adicional y opcional!):

Usar cross validation estratificada para determinar el valor de C óptimo

C: float, default=1.0
Inverse of regularization strength; must be a positive float. Like in support vector machines, smaller values specify stronger regularization.

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegressionCV.html

https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html#sklearn.model_selection.StratifiedKFold


### Ejercicio 2

---

### 2.1

In [29]:
data = pd.read_csv('../Data/Default.csv', sep="\t")

student_dummy = pd.get_dummies(data.student, drop_first = True, prefix = 'student')
student_dummy.head(3)

data = pd.concat([data, student_dummy], axis = 1)

Modificamos algunos datos para asignarle la etiqueta 'Unknown'

In [30]:
mask = data.default == 'No'
sample_count = round(np.sum(mask) * 0.2)
universe = np.reshape(np.where(mask), -1)
unknown_indexes = np.random.choice(universe, size = sample_count, replace = False)
data.loc[unknown_indexes, 'default'] = 'Unknown'
data.default.value_counts()

No         7734
Unknown    1933
Yes         333
Name: default, dtype: int64

### 2.2

In [31]:
X = data[['student_Yes', 'balance', 'income']]
y = data.default

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state = 127)

In [33]:
logistic_regression = LogisticRegression(penalty='none', multi_class = 'multinomial', class_weight = 'balanced')
logistic_regression.fit(X_train, y_train);

In [34]:
y_test_pred = logistic_regression.predict(X_test)
y_test_pred

array(['No', 'Yes', 'No', ..., 'Yes', 'No', 'Yes'], dtype=object)

In [35]:
accuracy_score(y_test, y_test_pred)

0.456

In [36]:
confusion_matrix(y_test, y_test_pred, )

array([[1000,  261,  673],
       [ 272,   63,  148],
       [   3,    3,   77]])

In [37]:
print(sum(y_test == 'No'))
print(sum(y_test == 'Unknown'))
print(sum(y_test == 'Yes'))

1934
483
83


Observemos que:

- Este clasificador es bastante malo.

- El código utilizado para ajustar un modelo de regresión logística multiclase es casi idéntico al de clasificación binaria, asignando el valor del parámetro `multi_class` al constructor `LogisticRegression`:

```python
logistic_regression = LogisticRegression(penalty='none', multi_class = 'multinomial', class_weight = 'balanced')
```
El valor balanced del parámetro `class_weight` da mayor peso a las instancias de categorías menos frecuentes.


### 2.3

Ahora vamos a construir un dataset balanceando las 3 categorías target del dataset original, y vamos a ajustar otra regresión logística

In [38]:
data.default.value_counts()

No         7734
Unknown    1933
Yes         333
Name: default, dtype: int64

In [39]:
data_unknown = data.loc[data.default == 'Unknown', ]
data_no = data.loc[data.default == 'No', ]
data_yes = data.loc[data.default == 'Yes', ]
data_balanced = pd.concat([data_unknown.sample(n = 333, replace = False), data_no.sample(n = 333, replace = False), data_yes], axis = 0)
data_balanced.default.value_counts()

Yes        333
No         333
Unknown    333
Name: default, dtype: int64

In [40]:
X = data_balanced[['student_Yes', 'balance', 'income']]
y = data_balanced.default

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state = 127)

In [42]:
logistic_regression = LogisticRegression(penalty='none', multi_class = 'multinomial')
logistic_regression.fit(X_train, y_train);

In [43]:
y_test_pred = logistic_regression.predict(X_test)


In [44]:
#logistic_regression.predict_proba(X_test)

In [45]:
accuracy_score(y_test, y_test_pred)

0.52

In [46]:
logistic_regression.classes_

array(['No', 'Unknown', 'Yes'], dtype=object)

In [47]:
confusion_matrix(y_test, y_test_pred, )

array([[51,  6, 26],
       [50,  6, 28],
       [ 9,  1, 73]])

Vemos que:

* el clasificador sigue teniendo una performance bastante mala, aunque es mejor que el obtenido en la solución anterior

* la clase peor clasificada en "Unknown"


<div class="div-dhds-fondo-1"> Conclusiones
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Conclusiones

---

- Modelo es útil para abordar problemas de clasificación con una variable target cualitativa o categórica, generalmente binaria.

- La relación entre la variable dependiente y los predictores es lineal al realizar la transformación logística de los datos.

- Pueden interpretarse los valores predichos por el modelo como "probabilidades" de cada uno de las categorías de la variable target.

- Podemos realizar la interpretación de la influencia de las variables predictoras en términos de odds-ratio.


<div class="div-dhds-fondo-1"> Referencias
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M4/CLASE_28_Regresion_Logistica/Presentacion/img/M4_CLASE_28_separador.png" align="center" />

</div>

## Referencias

---

<a href="https://towardsdatascience.com/animations-of-logistic-regression-with-python-31f8c9cb420" target= "_blank">Animations of logistic regression with python</a>

<a href="https://medium.com/swlh/from-animation-to-intuition-linear-regression-and-logistic-regression-f641a31e1caf" target= "_blank">From animation to intuition linear regression and logistic regression</a>

<a href="https://towardsdatascience.com/building-a-logistic-regression-in-python-step-by-step-becd4d56c9c8" target= "_blank">Building a logistic regression in python step by step</a>

<a href="https://dataaspirant.com/how-logistic-regression-model-works/" target= "_blank">How logistic regression model works?</a>

<a href="https://github.com/rasbt/python-machine-learning-book/blob/master/faq/regularized-logistic-regression-performance.md" target= "_blank">Regularization in logistic regression</a>

<a href="https://www.kaggle.com/satishgunjal/multiclass-logistic-regression-using-sklearn" target= "_blank">Multiclass logistic regression using sklearn</a>